File: 백준/Gold/15685. 드래곤 커브/드래곤 커브.py
# 15685번
N = int(input())
graph = [[0] * 101 for _ in range(101)]

# 방향 지정
# 만약 방향이 0이면 90도 회전하면 다음 방향은 1
# 만약 방향이 1이면 90도 회전하면 다음 방향은 2
# 만약 방향이 2이면 90도 회전하면 다음 방향은 3
# 만약 방향이 3이면 90도 회전하면 다음 방향은 0
# 즉 (방향 + 1)을 4로 나눈 나머지!

# 방향대로 움직이는 리스트 선언
dx = [0, -1, 0, 1]
dy = [1, 0, -1, 0]


for _ in range(N):
    y, x, d, g = map(int,input().split())
    # 방문한 곳 1로 처리
    graph[x][y] = 1

    # 커브 방향을 저장하는 리스트
    curve = [d]

    # 세대만큼 반복
    for i in range(g):
        for j in range(len(curve)-1, -1, -1):
            curve.append((curve[j] + 1) % 4)        # 다음 방향 지정

    # 드래곤 커브
    for i in curve:
        x += dx[i]
        y += dy[i]

        # 격자 밖으로 넘어가면 무시
        if x<0 or x>=101 or y<0 or y>=101:
            continue

        # 넘어가지 않았다면 방문처리
        graph[x][y] = 1

res = 0
for i in range(100):
    for j in range(100):
        # 1*1 모서리 확인
        if graph[i][j] and graph[i+1][j] and graph[i][j+1] and graph[i+1][j+1]:
            res += 1
print(res)

File: 백준/Gold/15685. 드래곤 커브/드래곤 커브.js
const { group } = require("console");
const fs = require("fs");
const inputData = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

// k세대 드래곤 커브는 k-1세대 드래곤 커브를 끝점을 기준으로 90도 시계 방향 회전 시킨 다음, 그것을 끝 점에 붙인 것
const n = +inputData[0];
const board = Array.from({ length: 101 }, () => Array(101).fill(0));

// x, y, d, g
// x와 y는 드래곤 커브의 시작 점, d는 시작 방향, g는 세대
// 0: x좌표가 증가하는 방향 (→) -> 90도 회전하면 1
// 1: y좌표가 감소하는 방향 (↑) -> 90도 회전하면 2
// 2: x좌표가 감소하는 방향 (←) -> 90도 회전하면 3
// 3: y좌표가 증가하는 방향 (↓) -> 90도 회전하면 0

const dx = [0, -1, 0, 1];
const dy = [1, 0, -1, 0];

for (let i = 0; i < n; i++) {
  let [y, x, d, g] = inputData[i + 1].split(" ").map(Number);
  board[x][y] = 1;

  // 커브 방향 저장
  const curve = [d];

  for (let j = 0; j < g; j++) {
    for (let k = curve.length - 1; k >= 0; k--) {
      curve.push((curve[k] + 1) % 4);
    }
  }

  // 보드판에 표시
  for (let j = 0; j < curve.length; j++) {
    x += dx[curve[j]];
    y += dy[curve[j]];

    if (x >= 0 && x <= 101 && y >= 0 && y <= 101) {
      board[x][y] = 1;
    }
  }
}

let res = 0;

for (let i = 0; i < 100; i++) {
  for (let j = 0; j < 100; j++) {
    if (
      board[i][j] &&
      board[i + 1][j] &&
      board[i][j + 1] &&
      board[i + 1][j + 1]
    ) {
      res += 1;
    }
  }
}

console.log(res);